File: packages/agent/src/engine.ts
import type { McpHandle } from "./mcp-client";
import type { ModelPreset, PresetName } from "./models";
import type { ChatProvider } from "./providers";
import { createAgentLoop, type AgentLoop, type AgentLoopOpts } from "./agent-loop";
import type { MemoryStore } from "./db/memory";
import type { SkillStore } from "./skills";
import type { Tracer } from "./tracing";

// Process-level hub for shared, expensive resources:
//   - three ChatProviders (DeepSeek for thinking-mode sessions, OpenAI for
//     non-thinking, Gemini — each on its own API key + rate-limit bucket)
//   - one MCP connection
//   - one Tracer (observability backend)
//   - the skill store and the agent-side memory KV
// Hands out AgentLoops on demand. Each AgentLoop has its own context
// buffer, system prompt and iteration budget but reuses these shared
// resources.
//
// Pure wiring: `createEngine` performs NO IO and reads NO env — clients,
// the MCP connection, tracer config and skill validation all happen in
// the composition root (supervisor main / a script's main), which passes
// the finished resources in. That keeps the full dependency graph visible
// in one place and makes the engine trivially mockable (it's an
// interface; tests pass plain objects).

// Provider kind + the ChatProvider wrapper live in ./providers. Re-export
// the kind for callers that only need the discriminator.
export type { ProviderKind } from "./providers";

export interface Engine {
  readonly mcp: McpHandle;
  readonly presets: Record<PresetName, ModelPreset>;
  // Engine-level meta-skill names loaded into every session unless the
  // session opts out (`includeEngineSkills: false`).
  readonly skills: readonly string[];
  readonly tracer: Tracer;
  readonly skillStore: SkillStore;
  readonly memory: MemoryStore;
  // Pick the provider wrapper based on the model name. The model name is
  // the source of truth — a loop resolves a preset name to a concrete
  // model at construction time; this method only routes that model to
  // its endpoint.
  resolveProvider(model: string): ChatProvider;
  startAgentLoop(opts: AgentLoopOpts): Promise<AgentLoop>;
  endAgentLoop(id: string): void;
  log(sessionId: string, ...parts: unknown[]): void;
  // Ends open loops, flushes the tracer, closes the MCP connection.
  shutdown(): Promise<void>;
}

export interface EngineDeps {
  providers: {
    deepseek: ChatProvider;
    openai: ChatProvider;
    gemini: ChatProvider;
  };
  mcp: McpHandle;
  presets: Record<PresetName, ModelPreset>;
  // Engine-level skills — loaded into every session this engine starts
  // (unless a session opts out via `includeEngineSkills: false`). Use for
  // meta-skills that apply across every domain — e.g. `routing` (when to
  // delegate to another skill). Per-session domain skills are passed via
  // `AgentLoopOpts.skills` instead.
  //
  // Resolved at `startAgentLoop` time, not engine-create time, so live
  // overlay edits (e.g. by the `dreaming` skill) take effect on the very
  // next session without an engine restart.
  skills?: string[];
  skillStore: SkillStore;
  memory: MemoryStore;
  tracer: Tracer;
}

export function createEngine(deps: EngineDeps): Engine {
  const { providers, mcp, presets, skillStore, memory, tracer } = deps;
  const engineSkills: readonly string[] = deps.skills ?? [];
  const agentLoops = new Map<string, AgentLoop>();

  function log(sessionId: string, ...parts: unknown[]): void {
    console.log(`[${new Date().toISOString()}]`, `[${sessionId}]`, ...parts);
  }

  const engine: Engine = {
    mcp,
    presets,
    skills: engineSkills,
    tracer,
    skillStore,
    memory,
    log,

    resolveProvider(model) {
      if (model.startsWith("deepseek")) return providers.deepseek;
      if (model.startsWith("gemini")) return providers.gemini;
      return providers.openai;
    },

    async startAgentLoop(opts) {
      if (agentLoops.has(opts.id)) {
        throw new Error(`agent-loop id ${opts.id} already exists`);
      }

      const sessionSkillNames = opts.skills ?? [];
      const includeEngineSkills = opts.includeEngineSkills ?? true;
      const engineSkillNames = includeEngineSkills ? engineSkills : [];

      // Resolve session-level skills first (required: missing one is a
      // signal-handling error). Then engine-level skills (best-effort:
      // missing meta-skill is logged and dropped so a typo in engine
      // config doesn't take down every session).
      //
      // Final iteration order (Object insertion order) is preserved:
      // session domain skills → engine meta-skills. The loop uses this
      // ordering when composing the actual system message. The union of
      // each skill's frontmatter `tools:` list defines what MCP tools
      // this session is allowed to see — synthetic agent-side tools
      // (set_memory, read_skill, invoke_sub_agent, ...) stay always-on.
      const resolvedSkills: Record<string, string> = {};
      const accumulated = new Set<string>();
      let wildcard = false;
      const mergeSkill = (skill: { body: string; tools: string[] | "*" }, name: string) => {
        resolvedSkills[name] = skill.body;
        if (skill.tools === "*") {
          wildcard = true;
        } else {
          for (const t of skill.tools) accumulated.add(t);
        }
      };
      for (const name of sessionSkillNames) {
        const skill = await skillStore.readSkill(name);
        if (skill === null) {
          throw new Error(
            `session skill "${name}" not found (skills/${name}.md and skills.default/${name}.md both missing)`,
          );
        }
        mergeSkill(skill, name);
      }
      for (const name of engineSkillNames) {
        const skill = await skillStore.readSkill(name);
        if (skill === null) {
          log(opts.id, `[warn] engine skill "${name}" not found, skipping`);
          continue;
        }
        mergeSkill(skill, name);
      }

      // Wildcard from ANY loaded skill collapses the union to "all MCP
      // tools" — expressed as a null allow-list (the loop treats null as
      // no filter).
      let allowedTools: Set<string> | null = wildcard ? null : accumulated;

      // Caller-side narrowing on top of the skill-derived set. The
      // workflow executor passes a step's `tools: [...]` whitelist this
      // way so the sub-agent sees an intersection of (skill says OK) ∩
      // (workflow says OK). null skill-side means wildcard → fall back
      // to the workflow's list verbatim.
      if (opts.toolWhitelist) {
        if (allowedTools === null) {
          allowedTools = new Set(opts.toolWhitelist);
        } else {
          allowedTools = new Set(
            [...allowedTools].filter((t) => opts.toolWhitelist!.has(t)),
          );
        }
      }

      const loop = createAgentLoop(engine, { ...opts, resolvedSkills, allowedTools });
      agentLoops.set(opts.id, loop);
      const skillsList = Object.keys(resolvedSkills).join(",");
      const toolsLabel = allowedTools === null ? "*" : String(allowedTools.size);
      log(
        opts.id,
        `agent-loop opened (preset=${loop.preset} → model=${loop.model}, effort=${loop.reasoningEffort}, skills=[${skillsList}], tools=${toolsLabel}${opts.parentId ? `, parent=${opts.parentId}` : ""})`,
      );
      return loop;
    },

    endAgentLoop(id) {
      const loop = agentLoops.get(id);
      if (!loop) return;
      loop.close();
      agentLoops.delete(id);
      log(id, "agent-loop closed");
    },

    async shutdown() {
      for (const id of [...agentLoops.keys()]) engine.endAgentLoop(id);
      // Flush buffered tracer events. Without this, traces from the final
      // session(s) before SIGTERM are silently dropped.
      await tracer.shutdown();
      await mcp.close();
    },
  };

  return engine;
}

File: agent-py/agent_py/engine.py
"""Engine — a process-level singleton owning the expensive shared resources:
LangChain models (cached per preset, built lazily) and one MCP connection. It
hands out sessions on demand.

Skill resolution — the union of frontmatter tools, wildcard handling, and the
intersection with a caller-supplied tool allow-list — lives here because it is
the domain logic that shapes every session.
"""

from __future__ import annotations

import json
from datetime import UTC, datetime

from langchain_core.language_models import BaseChatModel
from langchain_core.tools import BaseTool

from .mcp_client import McpHandle, connect_mcp
from .models import DEFAULT_PRESETS, ModelPreset, build_model
from .session import DEFAULT_MAX_ITERATIONS, Session
from .skills import read_skill, validate_all_skills
from .synthetic_tools import build_synthetic_tools


class Engine:
    def __init__(
        self,
        *,
        mcp: McpHandle,
        presets: dict[str, ModelPreset],
        engine_skills: list[str],
    ) -> None:
        self.mcp = mcp
        self.presets = presets
        self.engine_skills = engine_skills
        self._model_cache: dict[str, BaseChatModel] = {}

    # ─── models ───────────────────────────────────────────────────────

    def model_for_preset(self, preset_name: str) -> BaseChatModel:
        if preset_name not in self._model_cache:
            self._model_cache[preset_name] = build_model(self.presets[preset_name])
        return self._model_cache[preset_name]

    # ─── env ──────────────────────────────────────────────────────────

    async def read_timezone(self) -> str:
        try:
            raw = await self.mcp.call_tool("get_timezone", {})
            if raw.startswith("[tool error]"):
                return "UTC"
            return json.loads(raw).get("timezone", "UTC")
        except Exception:  # noqa: BLE001 — any failure → safe default
            return "UTC"

    # ─── sessions ─────────────────────────────────────────────────────

    def start_session(
        self,
        *,
        session_id: str,
        system_prompt: str | None = None,
        skill_names: list[str] | None = None,
        include_engine_skills: bool = True,
        preset: str = "base",
        max_iterations: int = DEFAULT_MAX_ITERATIONS,
        is_parent: bool = True,
        tool_whitelist: set[str] | None = None,
    ) -> Session:
        """Resolve skills, compute the tool allow-list, assemble the system
        prompt, and return a ready Session."""
        session_skill_names = skill_names or []
        engine_skill_names = self.engine_skills if include_engine_skills else []

        resolved_skills: dict[str, str] = {}
        accumulated: set[str] = set()
        wildcard = False

        def merge(name: str, *, required: bool) -> None:
            nonlocal wildcard
            skill = read_skill(name)
            if skill is None:
                if required:
                    raise ValueError(f'session skill "{name}" not found')
                self.log(session_id, f'[warn] engine skill "{name}" not found, skipping')
                return
            resolved_skills[name] = skill.body
            if skill.tools == "*":
                wildcard = True
            else:
                accumulated.update(skill.tools)

        for name in session_skill_names:
            merge(name, required=True)
        for name in engine_skill_names:
            merge(name, required=False)

        # A wildcard from any skill → None = "all MCP tools" (no filter).
        allowed_tools: set[str] | None = None if wildcard else accumulated

        # Caller-side narrowing on top of the skill-derived set (workflow
        # llm_agent passes a whitelist).
        if tool_whitelist is not None:
            if allowed_tools is None:
                allowed_tools = set(tool_whitelist)
            else:
                allowed_tools = {t for t in allowed_tools if t in tool_whitelist}

        # Filter MCP tools by the allow-list; synthetic tools are always on.
        if allowed_tools is None:
            mcp_tools: list[BaseTool] = list(self.mcp.tools)
        else:
            mcp_tools = [t for t in self.mcp.tools if t.name in allowed_tools]
        synthetic = build_synthetic_tools(self, is_parent=is_parent)
        tools = [*mcp_tools, *synthetic]

        # System prompt: caller prompt → each skill, joined by `---`.
        parts: list[str] = []
        if system_prompt:
            parts.append(system_prompt)
        parts.extend(resolved_skills.values())
        combined_system = "\n\n---\n\n".join(parts)

        preset_def = self.presets[preset]
        sess = Session(
            self,
            session_id=session_id,
            system_prompt=combined_system,
            tools=tools,
            model=self.model_for_preset(preset),
            max_iterations=max_iterations,
            preset_name=preset,
            model_name=preset_def.model,
        )
        tools_label = "*" if allowed_tools is None else str(len(allowed_tools))
        self.log(
            session_id,
            f"session opened (preset={preset} → model={preset_def.model}, "
            f"skills=[{','.join(resolved_skills)}], tools={tools_label})",
        )
        return sess

    # ─── misc ─────────────────────────────────────────────────────────

    def log(self, session_id: str, *parts: object) -> None:
        ts = datetime.now(UTC).isoformat()
        print(f"[{ts}] [{session_id}]", *parts)

    async def shutdown(self) -> None:
        await self.mcp.close()


async def create_engine(
    *,
    presets: dict[str, ModelPreset] | None = None,
    engine_skills: list[str] | None = None,
) -> Engine:
    mcp = await connect_mcp()

    # Validate all skills against the live MCP registry — fail early.
    validate_all_skills(mcp.tool_names)
    print(f"[engine] skill validation passed (mcp tools: {len(mcp.tool_names)})")

    return Engine(
        mcp=mcp,
        presets=presets or dict(DEFAULT_PRESETS),
        engine_skills=engine_skills or [],
    )
